src/cryptonote_core/cryptonote_core.cpp

```diff
@@ -1444,17 +1448,27 @@ namespace cryptonote
   //-----------------------------------------------------------------------------------------------
   void core::do_uptime_proof_call()
   {
-    std::vector<service_nodes::service_node_pubkey_info> states = get_service_node_list_state({ m_service_node_pubkey });
-
     // wait one block before starting uptime proofs.
+    std::vector<service_nodes::service_node_pubkey_info> const states = get_service_node_list_state({ m_service_node_pubkey });
     if (!states.empty() && states[0].info.registration_height + 1 < get_current_blockchain_height())
     {
-      m_submit_uptime_proof_interval.do_call(boost::bind(&core::submit_uptime_proof, this));
+      // If we haven't received our proof back from peers on the network, change ping granularity so
+      // that we resend after the acceptable window of time for other peers to accept our ping.
+      if (m_quorum_cop.get_uptime_proof(states[0].pubkey) == 0)
+      {
+        // NOTE: Need another once_a_time instance one because it is templated for absolutely no reason, Can't change the timing of the existing one
+        static epee::math_helper::once_a_time_seconds<UPTIME_PROOF_BUFFER_IN_SECONDS, true /*start_immediately*/> uptime_proof_not_known_interval;
+        uptime_proof_not_known_interval.do_call(boost::bind(&core::submit_uptime_proof, this));
```

Comment:
This second `once_a_time` is pretty gross (though really, as your comment alludes to, it's the horrible `once_a_time_seconds` class design that uses template parameters as a stand-in for constructor arguments that is really to blame here).  Appropriate outrage acknowledged, perhaps an alternative is to use just the main interval, but firing every UPTIME_PROOF_BUFFER_IN_SECONDS via a lambda like this instead of directly into `submit_uptime_proof`:

```C++
m_submit_uptime_proof_interval.do_call([&]() {
    uint64_t last_proof = m_quorum_cop.get_uptime_proof(states[0].pubkey);
    if (last_proof <= static_cast<uint64_t>(std::time() - UPTIME_PROOF_FREQUENCY_IN_SECONDS)) {
        submit_uptime_proof();
    }
});
```

Reply:
It's worth adding that this suggestion would also make it a little more robust: if an uptime proof sent after the normal 1h didn't come back it would try again at 1h5m, 1h10m, etc.

Reply:
I like this tweak. It works out pretty nicely as you said. Commit has been updated.